Challenge Submit 1

mobencm opened the pull request.

mobencm committed:
Challenge Submit

mobencm committed:
Add environment variable

mobencm committed:
add health check for pg DB

albertaouad commented:
@mobencm thank your for your submission, the team will start reviewing your PR in the short term and get back to you with some feedback, comments and questions. Please address them accordingly.

abed-sinno-byd commented on intermediate/challenge-1/app/core/config.py:
```diff
@@ -0,0 +1,63 @@
+    POSTGRES_DB: str
+    SQLALCHEMY_DATABASE_URI: Optional[PostgresDsn] = None
+
+    @validator("SQLALCHEMY_DATABASE_URI", pre=True)
```
I would choose not mix the validator functions within the class members, which keeps the classes consistent, and also makes reading or scanning the class members easier, more streamlined.

mobencm replied:
From the Pydantic documentation, to perform custom data validation of a model, we need to use a validator decorator around a class method.

abed-sinno-byd replied:
I meant to say that you could distribute the functions at the end of the class, below the variable definitions; similar to the grouping done in file *schemas/schedule.py*

mobencm replied:
Yes, I refactored this way in the new commit

abed-sinno-byd commented on intermediate/challenge-1/docker-compose.yml:
```diff
@@ -0,0 +1,36 @@
+      POSTGRES_USER: postgres
+      POSTGRES_DB: Audela
+      POSTGRES_PASSWORD: 123456789
+      POSTGRES_USER: root
```
`POSTGRES_USER` is defined twice here, perhaps you want to remove this line, and use 'root' as the value for POSTGRES_USER at line 18, while also using the 'root' using in `.env` .
As it is, the postgres failed to run for me with `docker-compose up` (the error being `FATAL:  role "postgres" does not exist`)

mobencm replied:
```suggestion : Sorry, I pushed the wrong docker-compose file
      --- 
services: 
  app: 
    build: 
      context: ./app
      dockerfile: Dockerfile
    command: "bash -c \"python initial_data.py && python main.py\"\n"
    depends_on: 
        postgres:
            condition: service_healthy
    links:
        - postgres
    env_file: 
      - .env
    ports: 
      - "8080:8080"
  postgres: 
    container_name: postgresDB
    environment:
      POSTGRES_USER: postgres
      POSTGRES_DB: Audela
      POSTGRES_PASSWORD: 123456789
    image: "postgres:latest"
    ports: 
      - "5555:5432"
    restart: always
    volumes: 
      - "./init-db/:/docker-entrypoint-initdb.d/"
      - "pg-data:/var/lib/postgresql/data"
    healthcheck:
      test: [ "CMD", "pg_isready", "-q", "-d", "${POSTGRES_USER}", "-U", "${POSTGRES_USER}" ]
      interval: 5s
      timeout: 3s
      retries: 5
version: "3"
volumes: 
  pg-data: ~


```

abed-sinno-byd commented on intermediate/challenge-1/app/api/api_v1/endpoints/login.py:
```diff
@@ -0,0 +1,35 @@
+    )
+    if not user:
+        raise HTTPException(status_code=400, detail="Incorrect email or password")
+    access_token_expires = timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
```
I believe you want to import from the `Settings` class, not `settings` module; fixing this could allow you to remove line 62 of config.py, which appears to be not needed or inappropriate.

Due to this apparently, I am unable to run the application, docker-compose returns the below:
```app_1       |   File "/usr/src/app/core/security.py", line 5, in <module>
app_1       |     from core.config import settings
app_1       |   File "/usr/src/app/core/config.py", line 7, in <module>
app_1       |     class Settings(BaseSettings):
app_1       |   File "/usr/src/app/core/config.py", line 35, in Settings
app_1       |     [str] = None
app_1       | TypeError: cannot unpack non-iterable NoneType object
```

While if I try running `main` from IDE, I am getting:
```
File "/Users/abdelrahmansinno/IdeaProjects/temp/amaze-us/intermediate/challenge-1/app/crud/crud_user.py", line 5, in <module>
    from core.security import get_password_hash, verify_password
  File "/Users/abdelrahmansinno/IdeaProjects/temp/amaze-us/intermediate/challenge-1/app/core/security.py", line 5, in <module>
    from core.config import settings
ImportError: cannot import name 'settings' from 'core.config' 
```

mobencm replied:
I don't know why docker-compose returns this error but I removed line 62 in config.py and imported the "settings" module and it worked fine.
for the case of the IDE, I think you should mark the app folder as the source root

abed-sinno-byd commented on intermediate/challenge-1/app/api/api_v1/endpoints/login.py:
```diff
@@ -0,0 +1,35 @@
+    )
+    if not user:
+        raise HTTPException(status_code=400, detail="Incorrect email or password")
+    access_token_expires = timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
```
I believe you want to import from the `Settings` class, not `settings` module; fixing this could allow you to remove line 62 of config.py, which appears to be not needed or inappropriate.

Due to this apparently, I am unable to run the application, docker-compose returns the below:
```app_1       |   File "/usr/src/app/core/security.py", line 5, in <module>
app_1       |     from core.config import settings
app_1       |   File "/usr/src/app/core/config.py", line 7, in <module>
app_1       |     class Settings(BaseSettings):
app_1       |   File "/usr/src/app/core/config.py", line 35, in Settings
app_1       |     [str] = None
app_1       | TypeError: cannot unpack non-iterable NoneType object
```

While if I try running `main` from IDE, I am getting:
```
File "/Users/abdelrahmansinno/IdeaProjects/temp/amaze-us/intermediate/challenge-1/app/crud/crud_user.py", line 5, in <module>
    from core.security import get_password_hash, verify_password
  File "/Users/abdelrahmansinno/IdeaProjects/temp/amaze-us/intermediate/challenge-1/app/core/security.py", line 5, in <module>
    from core.config import settings
ImportError: cannot import name 'settings' from 'core.config' 
```

abed-sinno-byd commented on intermediate/challenge-1/app/api/api_v1/endpoints/users.py:
```diff
@@ -0,0 +1,114 @@
+        *,
+        db: Session = Depends(deps.get_db),
+        password: str = Body(None),
+        full_name: str = Body(None),
```
Is there any reason for not using the UserUpdate object as input parameter instead, similarly to how it is used when updated a User by ID ?

mobencm replied:
UserUpdate object is used by PUT "/{user_id}" when a manager wants to change another user's information (e.g. upgrade a user's role from employee to manager), which does is not the case of the PUT "/me" which is simply invoked when the user wants to update his own information (name, password, email)

abed-sinno-byd commented on intermediate/challenge-1/.env:
```diff
@@ -0,0 +1,14 @@
+POSTGRES_USER=postgres
+POSTGRES_DB=Audela
+POSTGRES_PASSWORD=123456789
+FIRST_MANAGER=firsr_manager@audela.com
```
It seems more appropriate to avoid using the environment variables for storing the test-data for 'FIRST_MANAGER', instead these values can belong in Database Setup and/or test logic.

mobencm replied:
To keep sensitive data safe, I think to create the first user with a password (FIRST_MANAGER and FIRST_MANAGER_PASSWORD), it's better to use environment variable references instead of setting passwords directly in the configuration file or the database configuration. These are set by who’s responsible for the deployment environment.

for the EMAIL_TEST_USER variable which was needed for testing, yes it's more appropriate to separate it from the config.py file and provide it via a test config file with a database configuration in a controlled testing environment.

mobencm committed:
code refactoring after code review:
- config.py: Refactor  so as not to mix validation functions with class members
- conftest.py: Remove the EMAIL_TEST_USER environment variable replacing it with a random email
- Docker-compose : remove POSTGRES_USER duplicate definition

abdoulaye-byd commented on intermediate/challenge-1/app/api/api_v1/endpoints/users.py:
```diff
@@ -0,0 +1,114 @@
+        return user
+    if not crud.user.is_manager(current_user):
+        raise HTTPException(
+            status_code=400, detail="The user doesn't have enough privileges"
```
I think a `403` status code would be more suited here to signify a lack privilege for the given operation

abdoulaye-byd reviewed (commented):
Thank you for the submission. 

**Some of the highlights:**
- the codebase is neat and organized
- good use of dependency injection
- using background tasks to queue tasks for sending emails

**Some things that could be improved:**
- more consistency on HTTP status codes especially around Authentication/Authorization, there are places where `400` is returned where it should've been `403`
- `get_current_active_manager` could be dependency at the router level (see [docs](https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-in-path-operation-decorators/) )
- I understand it might be because of time constraints or simplicity sake, but the RBAC implementation could have been modeled around roles with their permissions, this will make for a design that's extensible should authz rules change. Also for the current implementation, `is_manager` could be an attribute of the jwt token, saving you a db query for every request (obviously the case where that field changes for given user must be taken into account)

albertaouad commented:
@mobencm thank you for your submission, we are closing this PR as we move to the next stage in the process.
